[Step5] 자동차 경주(리팩토링)

chomily opened the pull request:
안녕하세요! 5단계 미션 제출합니다 :)
피드백 부탁드려요 🙏

chomily committed:
update README.md

chomily committed:
refactor based on feedbacks

- delete unnecessary init block
- add input value at exception message
- replace mock with lambda

chomily committed:
refactor based on feedbacks

- add histories as Car constructor's second param

chomily committed:
refactor based on feedbacks

- add histories as Car constructor's second param

chomily committed:
refactor based on feedbacks

- refactor how to show car's location

chomily committed:
deep copy constructor parameters, add histories param at MoveHistory constructor

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/MoveHistory.kt:
```diff
@@ -2,8 +2,8 @@ package com.nextstep.racingcar.domain
 
-class MoveHistory {
-    private val histories = mutableListOf<Movement>()
+class MoveHistory(histories: MutableList<Movement> = mutableListOf()) {
```
처음 생성자에서 받을 때 부터 mutableList일 필요는 없어보여요.
내부에서 받은 list를 mutable로 변경하고 있으니, list를 받게 하는 것은 어떨까요?

chomily replied:
아하, 그러네요! List를 받도록 하면 불변 List나 가변 List 모두 파라미터로 받을 수 있겠군요! 수정하였습니다 :)

malibinYun commented on src/test/kotlin/com/nextstep/racingcar/domain/CarsTest.kt:
```diff
@@ -14,54 +12,31 @@ class CarsTest : StringSpec({
 
-        sujinCar.move(movingRule)
-        chomilyCar.move(notMovingRule)
+        sujinCar.move { MOVE }
+        chomilyCar.move { NONE }
```
Move를 직접 호출할 수도 있겠지만,
Car를 생성할 때, 움직였다는 내용을 담은 History객체를 넣어서 승리자에 대한 객체를 명시적으로 만들어볼 수 있겠네요 :)

chomily replied:
오호 수정이 덜 된 곳이 있었네요! 감사합니다 :) 수정했어요~~!

malibinYun reviewed (changes requested):
5단계 미션 정말 고생 많으셨어요!!
코드 변경점이 많지 않아서 크게 리뷰 드릴 게 없었네요 :)
몇 가지 코멘트 남겨두었으니 확인 부탁드릴게요!

이전에 반영해보지 못했던 피드백을 반영하거나, 시도해보고 싶은 점들이 있으면 반영 후에 다시 리뷰 요청 부탁드릴게요!
이만 미션을 마무리 하시고 싶으시다면 DM 혹은 코멘트 남겨두어주시면 미션 종료 하겠습니다 🙂

chomily committed:
refactor based on feedback

- change MoveHistory's constructor parameter Type(MutableList -> List)
- (test code)initialize car object with MoveHistory List to set test condition

chomily committed:
refactor test code

chomily commented:
이전 리뷰에서 반영 못했던 테스트 코드 리팩토링 추가했습니다:)
여기까지 하고 이번 미션은 마무리해도 좋을 것 같아요!
리뷰 감사합니다, 윤혁님! 🙏

malibinYun reviewed (approved):
자동차 경주 미션 완주 정말 고생 많으셨어요!!
수진님과 소통하며 리뷰한 시간이 즐거웠어요 :)
시간이 조금 들긴 했지만 미션을 끝까지 완주해주신 모습이 인상 깊었어요!!
다른 미션들도 꼭 완주하시길 응원합니다. 💪🔥
아무쪼록 이 미션을 하면서 고민했고, 알게 되었던 내용들이 앞으로 프로젝트를 진행하면서 작은 도움이 되었으면 좋겠어요 😊
